File: 0091-decode-ways/0091-decode-ways.java
class Solution {
    public int numDecodings(String s) {
        int n = s.length();
        // dp[j]: 截止到s[j-1]有多少种decode的方法，因为这题必须用到dp[i-2]，所以把j往后移一位
        int[] dp = new int[n + 1];
        dp[0] = 1;
        dp[1] = s.charAt(0) == '0' ? 0 : 1;
        for (int i = 2; i <= n; i++) {
            int first = Integer.parseInt(s.substring(i - 1, i));
            int second = Integer.parseInt(s.substring(i - 2, i));
            if (first > 0) { // 当前character不为0，那么它可以单独decode
                dp[i] += dp[i - 1]; // 其实就是dp[i] = dp[i - 1]，
            }
            if (second >= 10 && second <= 26) { // 保证了这个两位数的第一位不为0,如果valid这个二位数就可以单独decode
                dp[i] += dp[i - 2];
            }
        }
        return dp[n];
    }

}

File: 0091-decode-ways/0091-decode-ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        if s is None or s[0] == '0':
            return 0
        n = len(s)
        # dp[i] : the number of ways to decode s[:i]
        dp = [0] * (n + 1)
        dp[0] = 1
        dp[1] = 1

        for i in range(1, n):
            if int(s[i]) > 0:
                dp[i + 1] += dp[i]
            if s[i - 1] != '0' and 1 <= int(s[i - 1: i + 1]) <= 26:
                dp[i + 1] += dp[i - 1]
        
        print(dp)
        return dp[n]


